### PR TITLE
Make time zone parsing non-fallible

### PR DESCRIPTION
**Pull request by robertbastian**
Based on #5666 

Zones that are added to TZDB in the future should parse as `Etc/Unknown`. This means every unrecognized string should parse as `Etc/Unknown`. This then means that the `time_zone_id` field can be non-optional, as we can always use `Etc/Unknown` as the absent value.

The `metazone_id` field on the other hand need to be double optional: once for not-resolved, and once for no-metazone-for-zone. This change improves error cases in datetime formatting, i.e. `Some(None)` is a UTS-35 fallback, whereas `None` is a `MissingInputField`.

### CODE REVIEW

**Review by sffc — COMMENTED** [review]
I'm sorry, I don't follow the mental model here.

Time zone ID is not required, since the zone can be offset-only.

The `Option<Option<>>` metazone seems like it is an attempt to determine whether the MetazoneCalculator had been run yet. But, this is what the strongly typed time zone is supposed to enforce.

If you want to distinguish two different types of missing cases, "time zone ID not specified" and "time zone ID not found" seems like a more useful line.

Does this get us closer to the strongly typed time zone?

**Comment by robertbastian** [comment]
> Time zone ID is not required, since the zone can be offset-only.

An offset-only zone will report its time zone as `Etc/Unknown`. When formatted with a location-only format, it should format as `Unknown City Time`.

> The Option<Option<>> metazone seems like it is an attempt to determine whether the MetazoneCalculator had been run yet. But, this is what the strongly typed time zone is supposed to enforce.

This is a step towards strongly typed time zones. The resolved time zone will contain a non-optional variant, but still has to contain an optional metazone.

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
Review notes : code is fine, I don't understand the overall plan for the architecture here and defer to Shane on that.

**Comment by sffc** [comment]
I don't think we should necessarily make CustomTimeZone messy while we are still exploring the space, because it is the only way users currently have for formatting time zones in ICU4X and I consider icu_datetime mostly usable on main especially with the upcoming alpha release.

Right now CustomTimeZone is both "source input" and "calculated input". This PR puts it in a weird state where it seems like it wants to be "calculated input" but it is also still a "source input".

I'm also not fully in alignment with your Etc/Unknown mental model. I see there as being 3 states for the Time Zone ID field:

1. Not specified: `None`
2. Specified but not found: `Etc/Unknown`
3. Specified and found in data

It seems like you're trying to merge cases 1 and 2. It might make sense to do that on the "calculated input", but it doesn't really make sense on the "source input".

File: components/datetime/src/input.rs [hunk 1]
```diff
@@ -30,7 +30,7 @@ pub(crate) struct ExtractedInput {
     pub(crate) nanosecond: Option<NanoSecond>,
     pub(crate) offset: Option<UtcOffset>,
     pub(crate) time_zone_id: Option<TimeZoneBcp47Id>,
-    pub(crate) metazone_id: Option<MetazoneId>,
+    pub(crate) metazone_id: Option<Option<MetazoneId>>,
```
Comment (Manishearth) [inline components/datetime/src/input.rs:33]: nit: document the double option

**Comment by robertbastian** [comment]
> It seems like you're trying to merge cases 1 and 2. It might make sense to do that on the "calculated input", but it doesn't really make sense on the "source input".

I'm not sure what you mean by "calculated input" and "source input". UTS-35 does not really allow for absent time zone IDs, so any place where we keep the time zone ID optional, we're just later going to do `.unwrap_or(TimeZoneBcp47Id::unknown())`.

**Review by sffc — APPROVED** [review]
I now see how this is indeed a step toward https://github.com/unicode-org/icu4x/issues/5533#issuecomment-2412596448.

**Pull request merged**